Add ROI placement widget for scribbles in Slicer plugin

adamaji opened the pull request:
Took a crack at #433 as I'm working on some other ROI tools here. These changes add an ROI placement widget for scribbles in the Slicer plugin so the interface is more direct and there's no reliance on externally defined ROIs.

Right now the GUI looks like this for me on my windows machine as all I've done is add a new row for the markup placement widget. I think probably it would be suited better in the second row so there is more of a progression of operations from top to bottom, but looking for some input here. Perhaps some of the other UI elements could be rearranged as well.
![image](https://user-images.githubusercontent.com/3487395/169624551-3c79348e-e4b1-425b-8ee6-d3deccbf0f07.png)

Additionally the ROI display is hardcoded to be colored more neutral with an empty fill as described in https://github.com/Project-MONAI/MONAILabel/issues/422#issuecomment-927141895
![image](https://user-images.githubusercontent.com/3487395/169624449-28e3fbe3-c56e-4424-a563-a72bec3654e6.png)

cc @masadcv

SachidanandAlle commented:
This is great..  thanks for the contribution..
Code/logic  looks to me..
@masadcv can you please verify  the plugin at your end..

masadcv commented:
Many thanks! Will have a closer look at this tomorrow.

One minor point to better utilise widget space, can you move ROI widget to empty space next to Model: Histogram+Graphcut?

![169624551-3c79348e-e4b1-425b-8ee6-d3deccbf0f07](https://user-images.githubusercontent.com/3410225/169710670-77e54535-c880-4751-ac74-594ff277e4e6.png)

adamaji commented:
I've updated the UI so it looks like this in my environment:

![image](https://user-images.githubusercontent.com/3487395/169838943-c6abb85c-cc64-43d2-912d-a2e22447f129.png)

masadcv commented on plugins/slicer/MONAILabel/MONAILabel.py:
```diff
@@ -2084,6 +2093,21 @@ def updateBrushSize(self, value):
+        scribblesROINode.AddObserver(scribblesROINode.PointPositionDefinedEvent, self.onROIPointPlaced)
+        scribblesROINode.SetName("Scribbles ROI")
+        scribblesROINode.CreateDefaultDisplayNodes()
+        scribblesROINode.GetDisplayNode().SetFillOpacity(0.0)
```
Can you update this to have slight opacity to better differentiate inside and outside of the ROI?

e.g.
`scribblesROINode.GetDisplayNode().SetFillOpacity(0.4)` works well on my end.

masadcv commented on plugins/slicer/MONAILabel/MONAILabel.py:
```diff
@@ -1981,6 +1985,11 @@ def onResetScribbles(self):
 
+        # re-add ROI node if removed
+        if self.ui.scribblesPlaceWidget.currentNode() is None:
+            scribblesROINode = self.createScribblesROINode()
```
This is not the right place for re-initialising - please do it in the `onStartScribbling()` function or create a separate `initialiseROINodes()` function that gets called everytime a new sample is loaded...

masadcv commented on plugins/slicer/MONAILabel/MONAILabel.py:
```diff
@@ -2084,6 +2093,21 @@ def updateBrushSize(self, value):
 
+    def createScribblesROINode(self):
+        scribblesROINode = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLMarkupsROINode")
+        scribblesROINode.AddObserver(scribblesROINode.PointPositionDefinedEvent, self.onROIPointPlaced)
```
You are adding observers everytime this function is called, however I do not see a RemoveObserver logic anywhere? Maybe a good idea to cover that along with a `resetROI()` function....

masadcv replied:
New logic looks good to me..

masadcv commented on plugins/slicer/MONAILabel/MONAILabel.py:
```diff
@@ -1168,9 +1175,10 @@ def onNextSampleButton(self):
-            self.onResetScribbles()
             slicer.mrmlScene.Clear(0)
 
+        self.onResetScribbles()
```
I am not sure why you are moving this outside the if condition?
Is there a specific logic that requires this?

masadcv reviewed (changes requested):
Many thanks for looking into this. Overall, it looks quite good, however there are a few points to cover in my comments.

At the moment, it works fine if I add ROI with scribbles. However, without ROI, the scribbles seems to be broken. See this where I attempt this: https://www.youtube.com/watch?v=NuwwBTsWDrI  -  can you fix this issue? Possibly you need to return `None` for ROIs when no ROI is placed. This logic is currently missing which may be causing this.  

The second issue is missing `ResetROI()` functionality that needs to be utilized when loading new sample as well as changing labels in multilabel apps. Specifically the ROIs are not reset when new label is selected for multilabel app. See this example where I change label: https://www.youtube.com/watch?v=gGLLn9s_uKo - expected behaviour here is that the ROI be removed when I select kidney as my new label for scribbles.

Give me a ping once these are addressed, I will then give it another try! Let me know if anything is unclear...

masadcv commented on plugins/slicer/MONAILabel/Resources/UI/MONAILabel.ui:
```diff
@@ -645,6 +645,30 @@
+          <item>
+           <widget class="QLabel" name="scribblesROILabel">
+            <property name="text">
+             <string>Place ROI:</string>
```
Lets rename this to `ROI:` 

It may be understood that this is for placing ROI...

adamaji commented:
Thanks for the review @masadcv . I believe I've addressed all your comments with the latest push here. The scribbles ROI now acts more like the existing segmentation node, in that it only persists across one sample and is kept in a member variable. The issue with scribbles not working without an ROI was because the ROI node was defined, but returned bounds with lengths of 0, so added another check for control point definition there.

One thing I had trouble with was understanding some of the intended function behavior -- for example `clearScribbles` mentions that it is used for resetting tools to default but directly under that function is a `resetScribbles` function, so it's not clear to me how "resetting" scribbles is different from "clearing" here.

masadcv commented:
> Thanks for the review @masadcv . I believe I've addressed all your comments with the latest push here. The scribbles ROI now acts more like the existing segmentation node, in that it only persists across one sample and is kept in a member variable. The issue with scribbles not working without an ROI was because the ROI node was defined, but returned bounds with lengths of 0, so added another check for control point definition there.

Many thanks!! I will give this a go tomorrow morning and if all looks good, we can proceed with this PR... 

> One thing I had trouble with was understanding some of the intended function behavior -- for example `clearScribbles` mentions that it is used for resetting tools to default but directly under that function is a `resetScribbles` function, so it's not clear to me how "resetting" scribbles is different from "clearing" here.

“Resetting” functionality removes everything associated with scribbles setup (e.g. when you click next you can clear everything and start fresh)

"Clearing" functionality only clears the segmentation labels in scribbles label maps - while retaining all the nodes used by scribbles, this is used in multi-label case where after fine-tuning one label a user switches to a different label but still wants to work with scribbles on the same volume (hence no need to reinitialize these nodes)

I agree the comments may need to be updated and more descriptive in capturing these behaviors...

adamaji commented on plugins/slicer/MONAILabel/MONAILabel.py:
```diff
@@ -1907,15 +1928,10 @@ def getROIPointsXYZ(self, roiNode):
         roi_points_ras = [0.0] * 6
         if roiNode.__class__.__name__ == "vtkMRMLMarkupsROINode":
```
Makes sense, pushed an update

masadcv replied:
Thanks!

masadcv commented:
@adamaji 
Awesome! All looks good and works well on my setup.. lets wait for the checks to finish...

Thank you for looking into this and going through the review!

SachidanandAlle commented:
Adding ROI node is a cleaner solution rather than picking the first node added.
Thank you everyone for your contribution and making it better.
